src/chunk-stream.iced

```diff
@@ -20,28 +28,42 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
     # if we don't have a full block, dump everything into extra and skip this round
     if chunk.length < @block_size
       @extra = chunk
-    else
-      # if we have to, call the transform function multiple times until we have less than a block size remaining
-      if @exact_chunking
-        while chunk.length >= @block_size
-          @push(@transform_func(chunk[...@block_size]))
-          chunk = chunk[@block_size...]
-        @extra = chunk
-      # if the transform function can accept chunks of length block_size*n for n \in N, just make one transform function call
-      else
-        remainder = chunk.length % @block_size
-        # chop off extra
-        if remainder isnt 0
-          @extra = chunk[chunk.length-remainder...]
-          chunk = chunk[...chunk.length-remainder]
-        @push(@transform_func(chunk))
-
-    cb()
+      return cb(null, new Buffer(''))
+
+    # next, mangle the chunk so that it is an even multiple of the block size
+    remainder = chunk.length % @block_size
+    if remainder isnt 0
+      @extra = chunk[chunk.length-remainder...]
+      chunk = chunk[...chunk.length-remainder]
+
+    # finally, process the chunk (guaranteed to be the correct length)
+    esc = make_esc(cb, "Failed to transform the chunk")
+    await @_transform_chunk(chunk, esc(defer(out)))
+    cb(null, out)
+
+
+  # this is to be overridden by subclasses who want to output something else after the chunking is over
+  _flush_append : (cb) ->
+    return cb(null, null)
 
   _flush : (cb) ->
-    unless @writableObjectMode
-      while @exact_chunking and @extra and @extra.length >= @block_size
-        @push(@transform_func(@extra[...@block_size]))
-        @extra = @extra[@block_size...]
-      if @extra and @extra.length isnt 0 then @push(@transform_func(@extra))
-    cb()
+    esc = make_esc(cb, "Failed to flush")
+
+    # potentially deal with multiple flushed blocks
+    if @extra?.length >= @block_size
+      await @_transform_chunk(@extra, esc(defer(out_blocks)))
+      if out_blocks?
+        @push(out_blocks)
+
+    # either way, write out one final block (length guaranteed <= @block_size)
+    if @extra?.length isnt 0
+      await @transform_func(@extra, esc(defer(out_short)))
+      if out_short?
+        @push(out_short)
+
+    # allow subclasses to push something out at the very very end
+    await @_flush_append(esc(defer(out_flush)))
+    if out_flush?
```

Comment:
Ditto here. What does it mean to return null? If we're operating in object mode, and `null` is actually a valid output, is this a bug?